src/toil_scripts/lib/urls.py

```diff
@@ -0,0 +1,147 @@
+import base64
+import hashlib
+import os
+import subprocess
+from urlparse import urlparse
+
+
+def download_url(url, work_dir='.', name=None, s3_encryption_key_path=None):
+    """
+    Downloads URL, can pass in file://, http://, s3://, or ftp://
+
+    :param str url: URL to download from
+    :param str work_dir: Directory to download file to
+    :param str name: Name of output file, if None, basename of URL is used
+    :param str s3_encryption_key_path: Path to 32-byte encryption key if url points to S3 file that uses SSE-C
+    :return str: Path to the downloaded file
+    """
+    file_path = os.path.join(work_dir, name) if name else os.path.join(work_dir, os.path.basename(url))
+    if s3_encryption_key_path:
+        _download_encrypted_file(url, file_path, s3_encryption_key_path)
+    elif url.startswith('s3:'):
+        _download_s3_url(file_path, url)
+    else:
+        subprocess.check_call(['curl', '-fs', '--retry', '5', '--create-dir', url, '-o', file_path])
+    assert os.path.exists(file_path)
+    return file_path
+
+
+def download_url_job(job, url, name=None, s3_encryption_key_path=None):
+    """Job version of `download_url`"""
+    work_dir = job.fileStore.getLocalTempDir()
+    fpath = download_url(url, work_dir=work_dir, name=name, s3_encryption_key_path=s3_encryption_key_path)
+    return job.fileStore.writeGlobalFile(fpath)
+
+
+def s3am_upload(fpath, s3_dir, num_cores=1, s3_encryption_key_path=None):
+    """
+    Uploads a file to s3 via S3AM
+    For SSE-C encryption: provide a path to a 32-byte file
+
+    :param str fpath: Path to file to upload
+    :param str s3_dir: Ouptut S3 path. Format: s3://bucket/[directory]
+    :param int num_cores: Number of cores to use for up/download with S3AM
+    :param str s3_encryption_key_path: (OPTIONAL) Path to 32-byte key to be used for SSE-C encryption
+    """
+    assert s3_dir.startswith('s3://'), 'Format of s3_dir (s3://) is incorrect: {}'.format(s3_dir)
+    s3_dir = os.path.join(s3_dir, os.path.basename(fpath))
+    if s3_encryption_key_path:
+        base_url = 'https://s3-us-west-2.amazonaws.com/'
```

Comment:
The latest unstable release of s3am now has --sse-key-is-master which causes it to derive the per-file key for you. It does not hard-code the region but determines it automatically from the bucket. I would prefer to use that here and eliminate the derivation code, at least for this function.